# Part 1

In [3]:
#importing libraries 
import pandas as pd
import numpy as np
import os


In [5]:
# Define the path to the new dataset 
project_folder_path = r"C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data"

In [5]:
# Import customer.csv into a DataFrame Data Original
df_customers = pd.read_csv(r'C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data\Original Data\customers.csv', index_col = False)

In [7]:
# Import orders_products_combined.pkl into a DataFrame
df_ords_prods_merge = pd.read_pickle(r'C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data\Prepared Data\ords_prods_merge.pkl')

In [7]:
df_customers.head()

user_id First Name    Surnam  Gender       STATE  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   n_dependants fam_status  income  
0             3    married  165665  
1             0     single   59285  
2             2    married   99568  
3             0     single   42049  
4             1    married   40374

In [9]:
df_customers.describe()

user_id            Age   n_dependants         income
count  206209.000000  206209.000000  206209.000000  206209.000000
mean   103105.000000      49.501646       1.499823   94632.852548
std     59527.555167      18.480962       1.118433   42473.786988
min         1.000000      18.000000       0.000000   25903.000000
25%     51553.000000      33.000000       0.000000   59874.000000
50%    103105.000000      49.000000       1.000000   93547.000000
75%    154657.000000      66.000000       3.000000  124244.000000
max    206209.000000      81.000000       3.000000  593901.000000

In [8]:
# Rename columns
df_customers.rename(columns={
    'First Name': 'first_name',
    'Surnam': 'last_name',
    'STATE': 'state',
    'GENDER': 'gender',
    'AGE': 'age',
    'fam_status': 'family_status',
    'n_dependants': 'num_dependents'
}, inplace=True)

# Verify the changes
print(df_customers.head())

   user_id first_name last_name  Gender       state  Age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_dependents family_status  income  
0               3       married  165665  
1               0        single   59285  
2               2       married   99568  
3               0        single   42049  
4               1       married   40374  


In [11]:
# Strip leading/trailing spaces and convert column names to lower case
df_customers.columns = df_customers.columns.str.strip().str.lower()

In [23]:
print(df_customers.head())

   user_id first_name last_name  gender       state  age date_joined  \
0    26711    Deborah  Esquivel  Female    Missouri   48    1/1/2017   
1    33890   Patricia      Hart  Female  New Mexico   36    1/1/2017   
2    65803    Kenneth    Farley    Male       Idaho   35    1/1/2017   
3   125935   Michelle     Hicks  Female        Iowa   40    1/1/2017   
4   130797        Ann   Gilmore  Female    Maryland   26    1/1/2017   

   num_dependents family_status  income  
0               3       married  165665  
1               0        single   59285  
2               2       married   99568  
3               0        single   42049  
4               1       married   40374  


# I would leave all the columns, all of them provide relevant information for our analysis

# Check for mixed-type data:

In [13]:
# check for mixed-type data
for column in df_customers.columns:
    mixed_data = df_customers[column].apply(type).nunique() > 1
    if mixed_data:
        print(f"Column {column} has mixed data types.")

Column first_name has mixed data types.


# Check for Missing Values:

In [15]:
# Check for missing values in each column
missing_values = df_customers.isnull().sum()
print(missing_values)

# Check for percentage of missing values
missing_percentage = (df_customers.isnull().sum() / len(df_customers)) * 100
print(missing_percentage)

user_id               0
first_name        11259
last_name             0
gender                0
state                 0
age                   0
date_joined           0
num_dependents        0
family_status         0
income                0
dtype: int64
user_id           0.000000
first_name        5.459994
last_name         0.000000
gender            0.000000
state             0.000000
age               0.000000
date_joined       0.000000
num_dependents    0.000000
family_status     0.000000
income            0.000000
dtype: float64


# Handling Missing Values: The first_name column has 11259 (5.46%) missing values. They will be replaced by with 'Unknown'.

In [19]:
# Impute missing values in 'first_name' with 'Unknown'
df_customers['first_name'] = df_customers['first_name'].fillna('Unknown')


In [21]:
# Verify that there are no more missing values
missing_values_after_impute = df_customers.isnull().sum()
print(missing_values_after_impute)

user_id           0
first_name        0
last_name         0
gender            0
state             0
age               0
date_joined       0
num_dependents    0
family_status     0
income            0
dtype: int64


# Checking duplicate

In [23]:
# Display duplicate rows
duplicate_rows = df_customers[df_customers.duplicated()]
print(duplicate_rows)

Empty DataFrame
Columns: [user_id, first_name, last_name, gender, state, age, date_joined, num_dependents, family_status, income]
Index: []


# Checking Data Type Consistency

In [25]:
# Check data types
print(df_customers.dtypes)
print(df_ords_prods_merge.dtypes)

user_id            int64
first_name        object
last_name         object
gender            object
state             object
age                int64
date_joined       object
num_dependents     int64
family_status     object
income             int64
dtype: object
Unnamed: 0                   int64
order_id                     int64
user_id                      int64
order_number                 int64
order_dow                    int64
order_hour_of_day            int64
days_since_prior_order     float64
product_id                   int64
add_to_cart_order            int64
reordered                    int64
existing_merge            category
product_name                object
aisle_id                     int64
department_id                int64
prices                     float64
merge_flag                category
busiest_days                object
busiest_period_of_day       object
max_order                    int64
loyalty_flag                object
spending_flag               object
o

In [27]:
# Merge DataFrames
df_combined = pd.merge(df_ords_prods_merge, df_customers, on='user_id', how='left')

In [28]:

print(df_combined.shape)

(32433030, 31)


In [31]:
# Check for duplicates in the combined DataFrame
duplicates_combined = df_combined.duplicated().sum()
print(f'Number of duplicate rows in combined DataFrame: {duplicates_combined}')

Number of duplicate rows in combined DataFrame: 0


In [41]:
# Check for missing values in the combined DataFrame
missing_values_combined = df_combined.isnull().sum()
print("Missing values in combined DataFrame:")

Missing values in combined DataFrame:


In [ ]:
df_ords_prods_merge = df_ords_prods_merge.drop(columns=['Unnamed: 0'])

In [1]:
print(missing_values_combined)


NameError: name 'missing_values_combined' is not defined

In [45]:
df_combined.to_pickle(r"C:\Users\marci\15-07-2024Instacart Basket Analysis\02 Data\Prepared Data\Orders_products_all.pkl")